### PR TITLE
Builtin path, v2

### PR DESCRIPTION
**Pull request by faho**
## Description

This adds a `path` builtin to deal with paths.

It offers the following subcommands:

- `filter` to go through a list of paths and only print the ones that pass some filter - exist, are a directory, have read permission, ...
- `is` as a shortcut for `filter -q` to only return true if one of the paths passed the filter
- `basename`, `dirname` and `extension` to print certain parts of the path
- `change-extension` to change the extension to a different one (as a string operation)
- `normalize` and `resolve` to canonicalize the paths in various flavors
- `sort` to sort paths, also only using the basename or dirname as a key

The definition of "extension" here was carefully considered and should line up with how extensions are actually used - `~/.bashrc` doesn't have an extension, but `~/.conf.d` does (".d").

These subcommands all compose well - they can read from arguments or stdin (like string), they can use null-delimited input or output (input is autodetected - if a NULL happens in the first PATH_MAX bytes it switches automatically).

It is both a failglob exception (so like `set` if a glob passed to it fails it just doesn't get any arguments for it instead of triggering an error), and passes output to command substitution buffers explicitly split (like `string split0`) so newlines are easy to handle.

This PR is a rerun of #8265 because that was very long and hard to follow.

I intend to merge this for 3.5.0 in two weeks or so.

Fixes issue #7658

## TODOs:

- [X] Changes to fish usage are reflected in user documentation/manpages.
- [X] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst

There is one thing I'm not entirely sure about that would be awkward to change later:

- [ ] `filter` currently follows symlinks, so any filter applies to what the link points to.
      This isn't unreasonable, but it might be possible that turning it around and requiring an option to follow is the better choice. Or we leave it as-is and maybe add an option to ignore symlinks?

### CODE REVIEW

**Comment by krobelus** [comment]
BTW [here](https://lore.kernel.org/git/alpine.LFD.0.98.0705090825090.4062@woody.linux-foundation.org/) is a creative strategy for cleaning up an ugly sequence of commits but that's okay, I can look at the sum of the changes.

File: src/builtins/path.cpp [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    int n_transformed = 0;
+    arg_iterator_t aiter(argv, optind, streams, opts.null_in);
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
```
Comment (ridiculousfish) [inline src/builtins/path.cpp:674]: There's three important reasons for failure here:

1. `ENOENT/ENOTDIR`: Path does not exist
2. `EPERM`: Path may exist but you can't read it
3. `ELOOP`: Path is a symlink loop

For a symlink loop, it probably makes sense to "resolve as far as we can" as this does, but it would be good to add a test for that case since it's easily overlooked. Also we should change the language in the docs from "nonexistent paths" to "paths which do not exist, or otherwise cannot be resolved" or something, maybe you have ideas here.

File: doc_src/cmds/path.rst [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+::
+
+    path sort [(-z | --null-in)] [(-Z | --null-out)] \
+        [(-q | --quiet)] [(-v | --invert)] \
+        [--what=basename|dirname|path] [([PATH...]
+
+
+``path sort`` returns the given paths in sorted order. They are sorted in the same order as globs - alphabetically, but with runs of numerical digits compared numerically.
+
+With ``--invert`` or ``-v`` the sort is reversed.
```
Comment (ridiculousfish) [inline doc_src/cmds/path.rst:354]: I suggest `--reverse` / `-r` instead of `--invert`, to match sort(1).

File: src/builtins/path.cpp [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+static int path_sort(parser_t &parser, io_streams_t &streams, int argc, const wchar_t **argv) {
+    options_t opts;
+    opts.invert_valid = true;
+    opts.what_valid = true;
+    opts.unique_valid = true;
+    int optind;
+    int retval = parse_opts(&opts, &optind, 0, argc, argv, parser, streams);
+    if (retval != STATUS_CMD_OK) return retval;
+
+    auto func = +[] (const wcstring &x) {
```
Comment (ridiculousfish) [inline src/builtins/path.cpp:723]: LOL the `+ lambda` trick, first time encountering this 👏

File: src/builtins/path.cpp [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+    while (const wcstring *arg = aiter.nextstr()) {
+        auto real = wrealpath(*arg);
+
+        if (!real) {
+            // The path doesn't exist, so we go up until we find
+            // something that does.
+            wcstring next = *arg;
+            // First add $PWD if we're relative
+            if (!next.empty() && next[0] != L'/') {
+                next = wgetcwd() + L"/" + next;
```
Comment (ridiculousfish) [inline src/builtins/path.cpp:679]: This should use `parser.vars().get_pwd_slash()` for two reasons:

1. It allows sanely handling cases where `getcwd` fails (most importantly if $PWD has been unlinked)
2. It plays better with concurrent execution, where each thread might have its own CWD

Be aware that `get_pwd_slash()` may have symlinks.

Also note the `path_apply_working_directory` function, this may be as simple as

    wcstring next = path_apply_working_directory(*arg, parser.vars.get_pwd_slash());

File: src/builtins/path.cpp (hunk 3 above)
Comment (faho) [reply src/builtins/path.cpp:723]: Yeah, it's a workaround because lambda types are unique:

```c++
auto t = []{};
auto t2 = []{};
```

Here, t and t2 have different types. So reassigning breaks - `t = t2` complains about incompatible types. The `+` coerces it to a function pointer so it works (possibly at some slight cost, I dunno). It could also be possible to use std::function instead but that requires us to spell the type again, AFAICT.

See for example https://stackoverflow.com/questions/18889028/a-positive-lambda-what-sorcery-is-this.

File: src/builtins/path.cpp (hunk 4 above)
Comment (faho) [reply src/builtins/path.cpp]: As the kids don't say: Donespacito.

File: doc_src/cmds/path.rst [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+    path basename GENERAL_OPTIONS [PATH...]
+    path dirname GENERAL_OPTIONS  [PATH...]
+    path extension GENERAL_OPTIONS [PATH...]
+    path filter GENERAL_OPTIONS [(-v | --invert)] \
+        [-d] [-f] [-l] [-r] [-w] [-x] \
+        [(-t | --type) TYPE] [(-p | --perm) PERMISSION] [PATH...]
+    path is GENERAL_OPTIONS [(-v | --invert)] [(-t | --type) TYPE] \
+        [-d] [-f] [-l] [-r] [-w] [-x] \
+        [(-p | --perm) PERMISSION] [PATH...]
```
Comment (krobelus) [inline doc_src/cmds/path.rst:19]: no need for backslashes and "..." should have a leading space.  Also `[(-v | --invert)]` should be `[-v | --invert]` and options should probably be aligned in the source like in string.rst.

File: doc_src/cmds/path.rst [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+    path normalize GENERAL_OPTIONS [PATH...]
+    path resolve GENERAL_OPTIONS [PATH...]
+    path change-extension GENERAL_OPTIONS EXTENSION [PATH...]
+    path sort GENERAL_OPTIONS [(-v | --invert)] \
+        [-u | --unique] [--key=basename|dirname|path] [([PATH...]
+
+    GENERAL_OPTIONS := [(-z | --null-in)] [(-Z | --null-out)] [(-q | --quiet)]
```
Comment (krobelus) [inline doc_src/cmds/path.rst:26]: Are we planning to add null switches to other commands as well? I don't see `path` and `string` being all that different in this regard.

File: doc_src/cmds/path.rst [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+It returns 0 if any normalization or resolution was done, i.e. any given path wasn't in canonical form.
+
+Examples
+^^^^^^^^
+
+::
+
+   >_ path resolve /bin//sh
+   # The "//" is squashed, and /bin is resolved if your system links it to /usr/bin.
+   # sh here is bash (on an Archlinux system)
```
Comment (krobelus) [inline doc_src/cmds/path.rst:291]:
```suggestion
+   # sh here is bash (on an Arch Linux system)
```

File: share/completions/path.fish [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a basename -d 'Give basename for given paths'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a dirname -d 'Give dirname for given paths'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a extension -d 'Give extension for given paths'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a change-extension -d 'Change extension for given paths'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a normalize -d 'Normalize given paths (remove ./, resolve ../ against other components..)'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a resolve -d 'Normalize given paths and resolve symlinks'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a filter -d 'Print paths that match a filter'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a is -d 'Return true if any path matched a filter'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a match -d 'Match paths against a glob'
+complete -f -c path -n "test (count (commandline -opc)) -lt 2" -a expand -d 'Expand globs'
```
Comment (krobelus) [inline share/completions/path.fish:13]: I think you forgot "sort"

File: src/builtins/path.cpp [hunk 9]
```diff
@@ -0,0 +1,895 @@
...
+    UNUSED(streams);
+    opts->have_key = true;
+    opts->key = w.woptarg;
+    return STATUS_CMD_OK;
+}
+
+/// This constructs the wgetopt() short options string based on which arguments are valid for the
+/// subcommand. We have to do this because many short flags have multiple meanings and may or may
+/// not require an argument depending on the meaning.
+static wcstring construct_short_opts(options_t *opts) {  //!OCLINT(high npath complexity)
```
Comment (krobelus) [inline src/builtins/path.cpp:378]: I think we can drop this OCLINT comment

File: src/builtins/path.cpp [hunk 10]
```diff
@@ -0,0 +1,895 @@
...
+    if (retval != STATUS_CMD_OK) return retval;
+
+    auto func = +[] (const wcstring &x) {
+        return wbasename(x);
+    };
+    if (opts.have_key) {
+        if (std::wcscmp(opts.key, L"basename") == 0) {
+            // Do nothing, this is the default
+        } else if (std::wcscmp(opts.key, L"dirname") == 0) {
+            func = +[] (const wcstring &x) {
```
Comment (krobelus) [inline src/builtins/path.cpp:733]: I guess we don't need the + here. Also maybe call it something like `keyfunc`?

**Comment by krobelus** [comment]
> `filter` currently follows symlinks, so any filter applies to what the link points to.

Right, if /bin is a symlink then both `path filter --type=dir /bin` and `path filter --type=linke /bin` match.
This is a bit weird but it's probably the correct behavior in most situations, so it sounds ok.

**Comment by krobelus** [comment]
I took a tired look at everything, looks fine

File: doc_src/cmds/path.rst (hunk 6 above)
Comment (faho) [reply doc_src/cmds/path.rst:26]: I believe we should add them to string as well, yes.

The main reason this is in path first is because it's *necessary* for path to be complete - NUL is the only byte you can use to separate paths unambiguously, and so it exists in a lot of commands (`find -print0`). Whereas string already passes through NULs from stdin, and has `string split0` that you can use to *work* with NUL-delimited strings by passing them as arguments.

Unfortunately `string` won't be able to use the PATH_MAX trick, because strings are arbitrary.

File: src/builtins/path.cpp (hunk 10 above)
Comment (faho) [reply src/builtins/path.cpp]: Renamed to keyfunc (and "funced" to "key", because that's what it does - it gets the key to a value).

I don't want to drop the "+" here because that leaves us with a weird asymmetry.

File: src/builtins/path.cpp (hunk 9 above)
Comment (faho) [reply src/builtins/path.cpp:378]: I'm not sure if oclint is still a thing. I couldn't get it to run last time I tried.

If it's not, we should remove all of these.

File: share/completions/path.fish (hunk 8 above)
Comment (faho) [reply share/completions/path.fish]: Yeah, the completions needed an update.

File: src/builtins/path.cpp (hunk 10 above)
Comment (krobelus) [reply src/builtins/path.cpp]: > I don't want to drop the "+" here because that leaves us with a weird asymmetry.

Sure. IMO the asymmetry is justified because `+` ist just a short way of spelling a declaration's type.
When I first tried to figure out what the `+` means I removed it and recompiled, but this one actually makes no difference which makes Occam's razor less sharp.

**Comment by faho** [comment]
It is done. Merged.
